basicdataset/src/main/java/ai/djl/basicdataset/nlp/PennTreebank.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.basicdataset.nlp;
+
+import ai.djl.Application;
+import ai.djl.basicdataset.BasicDatasets;
+import ai.djl.basicdataset.RawDataset;
+import ai.djl.ndarray.NDManager;
+import ai.djl.repository.Artifact;
+import ai.djl.repository.MRL;
+import ai.djl.repository.Repository;
+import ai.djl.training.dataset.Batch;
+import ai.djl.training.dataset.Dataset;
+import ai.djl.translate.TranslateException;
+import ai.djl.util.Progress;
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * The Penn Treebank (PTB) project selected 2,499 stories from a three year Wall Street Journal
+ * (WSJ) collection of 98,732 stories for syntactic annotation.
+ */
+public class PennTreebank implements RawDataset<Path> {
```

Comment:
Maybe rename this to PennTreebankText as it is not the full penn treebank due to lacking labels. We don't want to have users come to this later and be confused as to what it is. In that case, it is better to use the name to help clarify what it is, and also probably use documentation to help add additional details to the explanation.

Can you also make this implement TextDataset instead of (or in addition to) RawDataset? The RawDataset was mostly intended as a hack for FastText, which doesn't support training through the standard DJL API. Here, we do want normal DJL models to work with it. The dataset you are working with is one that features unlabeled data organized in sentences where each sentence is on a separate line. So, I recommend having the dataset treat each line as a record in the dataset.

There are different things you do with these kinds of datasets. You could use it to construct word embeddings like [word2vec](https://d2l.djl.ai/chapter_natural-language-processing-pretraining/word2vec.html). Another common use case would be [NLP Pretraining](https://d2l.djl.ai/chapter_natural-language-processing-pretraining/index.html) like you would use for BERT-style transformer models. The goal with this dataset is to make it easy for these use cases.

You will also run into the problem where the DJL [`Record`](https://javadoc.io/doc/ai.djl/api/latest/ai/djl/training/dataset/Record.html) will expect some kind of label for the dataset, but this dataset does not have labels. In that situation, you can use either an empty NDList or null for the label.

Reply:
OK, thanks, it helps a lot. I’ll change my codes.